python/lsst/daf/butler/_butler.py

```diff
@@ -333,10 +333,13 @@ def makeRepo(root: str, config: Union[Config, str, None] = None, standalone: boo
             if not os.path.isdir(uri.ospath):
                 safeMakeDir(uri.ospath)
         elif uri.scheme == "s3":
-            s3 = boto3.resource("s3")
-            # implies bucket exists, if not another level of checks
-            bucket = s3.Bucket(uri.netloc)
-            bucket.put_object(Bucket=uri.netloc, Key=uri.relativeToPathRoot)
+            # implies bucket already exists
+            s3 = boto3.client("s3")
+            if uri.relativeToPathRoot != ".":
+                # to be dir-equivalent, S3 this requires URIs ending on '/'
+                if not uri.geturl().endswith('/'):
+                    uri.updateFile(uri.relativeToPathRoot+'/')
+                s3.put_object(Bucket=uri.netloc, Key=uri.relativeToPathRoot+'/')
```

Comment:
Since you now have to add the `/` again here what is the point of the previous call to `updateFile`?

Reply:
`updateFile` will update and remember the addition of `/` to the uri, this manipulates the `uri.path` attribute. The `uri.relativeToPathRoot` strips the left hand side separator again so i have to add it again.

I can not use the `uri.path` option because the named directory will be precluded by a `/`. If I try inserting this what I will get is a key named `/` under which I will have a key named `dir`. In addition if you try to inset keys starting with `/` from the web interface to S3 it will complain that it's not allowed. 

The following, hopefully, demonstrates this better than I can put it into words: 

```import lsst.daf.butler as dafButler
uri = dafButler.ButlerURI("s3://bucket/dir")
print(uri)
print(uri.relativeToPathRoot)
print(uri.path)
```
outputs:
```
's3://bucket/dir'
'dir'
'/dir'
```
After updating the `path` attribute via `updateFile:
```
uri.updateFile(uri.relativeToPathRoot+'/')
print(uri)
print(uri.relativeToPathRoot)
print(uri.path)
```
outputs
```
's3://bucket/dir/'
'dir'
'/dir/'
```

The end conclusion is that I need to know in advance whether I'm pointing to a directory, in which case I need no leading separators but I do require a trailing separator, or if I'm pointing to a file in which case I don't need any separators. I update the URI to append the separator in case that at some point later anyone uses that URI to construct other URIs they don't get burned when they use `updateFile` to find out their root directory in the URI disappeared. My thought process was that this is one of the rare places where it's safe to acknowledge that we expect a directory so ensuring that is explicitly written to the URI could not hurt. 

To also be explicit about another point: I do not need that `s3.put_object` call there explicitly. As long as that base URI is updated appropriately, the root directory will implicitly be created during first `butler.put` call. That will call `S3Datastore.put` which will put some relative path of the form `root/possibly_yet_another_dir/filename.ext` when all 3 keys will be created: 2 directories `root/` and `root/possibly_yet_another_dir/` and the file key `root/possibly_yet_another_dir/filename.ext`. This actually should occur *almost* every time because *almost* every time there is a `root/butler.yaml` key being inserted. I was afraid of the `outfile` parameter however so I didn't dare to leave it to chance.

Reply:
Any time I see a `/` being added explicitly to a path it indicates a hack.  There are two things we know from makeRepo:

1. The URI must refer to a "directory" and not a file.
2. You don't know if the butler.yaml will be written to that location but the root URI will be rewritten into that config if it's relocated.

So it seems to me that changing the ButlerURI code to support something like:

```python
uri = ButlerURI(root, forceDirectory=True)
```
would fix your problems?  If you say it's a directory we always add a `/` to the end of the string in the constructor if there isn't one there already and you get exactly what you are doing her but without the hack. You also guarantee a "directory" like URI is written to the output config if the output config is written somewhere else.  You can also ensure that `relativeToPathRoot` always includes a trailing slash since the URI will know it's a directory URI (if forceDirectory is false (the default) it means you can't tell -- you can still set the isDirectory internal flag if you see a trailing slash).

On a related note I think the reason you can't get `<butlerRoot>` to work properly is that you aren't setting `self.configFile` in `__initFromS3YamlFile`.

Reply:
Sorry it took me a while, I was swamped with Uni-related work and just couldn't catch up. I made the fixes requested. Hope you like the changes. 

Specifics I'm concerned about: 
1) did I guess all the places where it's safe to assume that I can `forceDirectory` 
2) do you like the `Butler.makeRepo` better like this 

I changed test_location.py because I had a hard time keeping up with what is testing what and whether all cases are covered. I hope you don't mind the re-arrangement of the test data. Mostly from what I've seen it's all my mess originally anyhow. All the same original tests are kept except 1 and some new ones added. I can revert if required.